armi/reactor/converters/axialExpansion/axialExpansionChanger.py

```diff
@@ -0,0 +1,396 @@
+    referenceAssembly: :py:class:`Assembly <armi.reactor.assemblies.Assembly>`, optional
+        Assembly whose mesh other meshes will conform to if isDetailedAxialExpansion is False.
+        If not provided, will assume the finest mesh assembly which is typically fuel.
+
+    Notes
+    -----
+    Calling this method will result in an increase in mass via applyColdHeightMassIncrease!
+
+    See Also
+    --------
+    :py:meth:`armi.reactor.converters.axialExpansionChanger.AxialExpansionChanger.applyColdHeightMassIncrease`
+    """
+    assems = list(assems)
+    if not referenceAssembly:
+        referenceAssembly = _getDefaultReferenceAssem(assems)
+    axialExpChanger = AxialExpansionChanger(isDetailedAxialExpansion)
+    for a in assems:
+        axialExpChanger.setAssembly(a, expandFromTinputToThot=True)
+        axialExpChanger.applyColdHeightMassIncrease()
+        axialExpChanger.expansionData.computeThermalExpansionFactors()
+        axialExpChanger.axiallyExpandAssembly()
+    if not isDetailedAxialExpansion:
+        for a in assems:
+            a.setBlockMesh(referenceAssembly.getAxialMesh())
+    # update block BOL heights to reflect hot heights
+    for a in assems:
+        for b in a:
+            b.p.heightBOL = b.getHeight()
+            b.completeInitialLoading()
+
+
+class AxialExpansionChanger:
+    """
+    Axially expand or contract assemblies or an entire core.
+
+    Attributes
+    ----------
+    linked : :py:class:`AssemblyAxialLinkage`
+        establishes object containing axial linkage information
+    expansionData : :py:class:`ExpansionData <armi.reactor.converters.axialExpansionChanger.ExpansionData>`
+        establishes object to store and access relevant expansion data
+
+    Notes
+    -----
+    - Is designed to work with general, vertically oriented, pin-type assembly designs. It is not set up to account
+      for any other assembly type.
+    - Useful for fuel performance, thermal expansion, reactivity coefficients, etc.
+    """
+
+    def __init__(self, detailedAxialExpansion: bool = False):
+        """
+        Build an axial expansion converter.
+
+        Parameters
+        ----------
+        detailedAxialExpansion : bool, optional
+            A boolean to indicate whether or not detailedAxialExpansion is to be utilized.
+        """
+        self._detailedAxialExpansion = detailedAxialExpansion
+        self.linked = None
+        self.expansionData = None
+
+    def performPrescribedAxialExpansion(
+        self, a, componentLst: list, percents: list, setFuel=True
+    ):
+        """Perform axial expansion/contraction of an assembly given prescribed expansion percentages.
+
+        .. impl:: Perform expansion/contraction, given a list of components and expansion coefficients.
+            :id: I_ARMI_AXIAL_EXP_PRESC
+            :implements: R_ARMI_AXIAL_EXP_PRESC
+
+            This method performs component-wise axial expansion for an Assembly given expansion coefficients
```

Comment:
For now, ARMI ignores `ruff` rule D404:

> First word of the docstring should not be "This"

But only because I have been too busy to fix it in dozens of places. It's a good rule of thumb to not though, because nearly everything this number two is more active:

1. This method performs axial expansion.
2. Perform axial expansion.